macOS: Undo/Redo for changes to windows, tabs, and splits

mitchellh opened the pull request:
This PR implements the ability to undo/redo new and closed windows, tabs, and splits.

## Demo

https://github.com/user-attachments/assets/98601810-71b8-4adb-bfa4-bdfaa2526dc6

## Details

### Undo Timeout

Running terminal sessions _remain running_ for a configurable period of time after close, during which time they're undoable. This is similar to "email unsend" (since email in the traditional sense can't be unsent, clients simply hold onto it for a period of time before sending).

This behavior is not unique to Ghostty. The first and only place I've seen it is in iTerm2. And iTerm2 behaves similarly, although details of our behavior and our implementation vary greatly. 

The configurable period of time is done via the `undo-timeout` configuration. The default value is 5 seconds. This feels reasonable to be and is grounded in being the default for iTerm2 as well, so it's probably a safe choice.

Undo can be disabled by setting `undo-timeout = 0`. 

### Future

The actions that can be potentially undone/redone can be easily expanded in the future. Some thoughts on things that make sense to me:

- Any sort of split resizing, including equalization
- Moving tabs or splits

#### What about Linux?

I'd love to support this on Linux. I don't think any other terminal on Linux has this kind of feature (definitely might be wrong, but I've never seen it and I've looked at a lot of terminal emulators 😄 ). But there's some work to be done to get there.

## TODO for the Draft PR

This is still a draft. There are some items remaining (list will update as I go):

- [x] Undoing a closed window is sometimes buggy still and I'm not sure why, I have to dig into this.
- [x] New window should be undoable
- [x] New tab should be undoable
- [x] Close All Windows should be undoable
- [x] I think I have to get rid of TerminalManager. Undone windows won't be in TerminalManager's list of controllers and I think that's going to break a lot of things.
- [x] I haven't tested this with QuickTerminal at all. I expect bugs there but I want undo to work with splits there.
- [x] Close window with the red traffic light button doesn't trigger undo
- [x] Closing window with multiple tabs undoes them as separate windows

mitchellh committed:
wip: undo

mitchellh committed:
macos: implement a custom ExpiringUndoManager, setup undo for new/close

mitchellh committed:
macos: setup undo responders at the AppDelegate level

mitchellh committed:
macos: basic undo close window, not very robust yet

mitchellh committed:
macos: basic undo tab, not quite working

mitchellh committed:
macos: more robust undo tab that goes back to the same position

mitchellh committed:
add undo/redo keybindings, default them on macOS

mitchellh committed:
macos: fix an incorrect bindable write during view update

mitchellh committed:
macos: configurable undo timeout

mitchellh committed:
macos: remove debug log

mitchellh committed:
apprt/glfw,gtk: noop undo/redo actions

mitchellh committed:
macos: compile errors in CI

mitchellh committed:
Make undo/redo app-targeted so it works with no windows

mitchellh committed:
macos: remove TerminalManager

All logic related to TerminalController is now in TerminalController.

mitchellh committed:
macos: undo new window

mitchellh committed:
macos: initial window shouldn't support undo

mitchellh committed:
macos: undo new tab

mitchellh committed:
macos: undo close all windows

leonjza commented:
Cool feature, thank you! Curious if you'd consider making the keybind to restore cmd+shift+t rather than cmd+z, mimicking the behavior of browsers and allowing muscle memory to be transferred.

mitchellh committed:
config: add super+shift+t as a default undo too to mimic browsers

mitchellh committed:
macos: only process reopen if already activated

mitchellh committed:
macos: make sure we're not registering unnecessary undos

mitchellh committed:
macos: fix some CI build issues

mitchellh commented:
> Cool feature, thank you! Maybe a bit tongue in cheek, but curious if you'd consider making the key bind to restore cmd+shift+t rather than cmd+z, mimicking the behavior of browsers and allowing muscle memory to be transferred?

Good idea. Done.

mitchellh committed:
macos: address quick terminal basic functionality with new API

mitchellh committed:
macos: show quick terminal on undo/redo

mitchellh commented:
Alright, all known TODOs addressed. I think I ended up fixing a handful of quick terminal issues here not related to this PR at all. But the good news is undo works great with the quick terminal too now. 😄 

I'll keep letting this sit a bit longer and consider what else I may have missed, but this is looking close.

mitchellh commented:
Flipping this to ready for review. I think any other changes here will be minor.

mitchellh committed:
macos: red traffic light should be undoable

mitchellh committed:
macos: fix undo/redo for closing windows with multiple tabs

When closing a window that contains multiple tabs, the undo operation
now properly restores all tabs as a single tabbed window rather than
just restoring the active tab.

The implementation:
- Collects undo states from all windows in the tab group before closing
- Sorts them by their original tab index to preserve order
- Clears tab group references to avoid referencing garbage collected objects
- Restores all windows and re-adds them as tabs to the first window
- Tracks and restores which tab was focused (or focuses the last tab if none were)

AI prompts that generated this commit are below.

Each separate prompt is separated by a blank line, so this session was
made up with many prompts in a back-and-forth conversation.

> We need to update the undo/redo implementation in
> @macos/Sources/Features/Terminal/TerminalController.swift `closeWindowImmediately`
> to handle the case that multiple windows in a tab group are closed all at once,
> and to restore them as a tabbed window. To do this, I think we should collect
> all the `undoStates`, sort them by `tabIndex` (null at the end), and then on j
> restore, restore them one at a time but add them back to the same tabGroup. We
> can't use the tab group in the `undoState` because it will be garbage collected
> by then. To be sure, we should just set it to nil.

I should note at this point that the feature already worked, but the
code quality and organization wasn't up to my standards. If someone
using AI were just trying to make something work, they might be done at
this point.

I do think this is the biggest gap I worry about with AI-assisted
development: bridging between the "it works" stage at a junior quality
and the "it works and is maintainable" stage at a senior quality. I
suspect this will be a balance of LLMs getting better but also senior
code reviewers remaining highly involved in the process.

> Let's extract all the work you just did into a dedicated private method
> called `registerUndoForCloseWindow`

Manual: made some tweaks to comments, moved some lines around, didn’t change
any logic.

> I think we can pull the tabIndex directly from the undoState instead of
> storing it in a tuple.

> Instead of `var undoStates`, I think we can create a `let undoStates` and
> build and filter and sort them all in a chain of functional mappings.

> Okay, looking at your logic for restoration, the `var firstController` and
> conditionals are littly messy. Can you make your own pass at cleaning those
> up and I'll review and provide more specific guidance after.

> Excellent. Perfect. The last thing we're missing is restoring the proper
> focused window of the tab group. We should store that and make sure the
> proper window is made key. If no windows were key, then we should make the
> last one key.

> Excellent. Any more cleanups or comments you'd recommend in the places you
> changed?

Notes on the last one: it gave me a bunch of suggestions, I rejected most but
did accept some.

> Can you write me a commit message summarizing the changes?

It wrote me a part of the commit message you're reading now, but I
always manually tweak the commit message and add my own flair.

mitchellh committed:
macos: clear out the surface trees to prevent repeat undo

see the comment
